packages/codemirror-extension/src/index.ts

```diff
@@ -233,13 +277,70 @@ function activateEditorCommands(
     }
   });
 
+  /**
+   * Handle the settings of new widgets.
+   */
+  notebookTracker.widgetAdded.connect((sender, widget) => {
+    widget.content.widgets.forEach(cell => {
+      if (cell.inputArea.editor instanceof CodeMirrorEditor) {
+        const cm = cell.inputArea.editor.editor;
+        // Do not set scrollPastEnd option.
+        cm.setOption('keyMap', keyMap);
+        cm.setOption('theme', theme);
+        cm.setOption('styleActiveLine', styleActiveLine);
+        cm.setOption('styleSelectedText', styleSelectedText);
+        cm.setOption('selectionPointer', selectionPointer);
+        cm.setOption('lineWiseCopyCut', lineWiseCopyCut);
+      }
+    });
+  });
+
+  notebookTracker.activeCellChanged.connect(onActiveCellChanged, this);
+
+  let activeCell: Cell | null = null;
+
+  commands.addCommand('codemirror:leave-vim-insert-mode', {
+    label: 'Leave VIM Insert Mode',
+    execute: args => {
+      if (activeCell) {
+        let editor = activeCell.editor as CodeMirrorEditor;
+        (CodeMirror as any).Vim.handleKey(editor.editor, '<Esc>');
+      }
+    },
+    isEnabled
+  });
+
+  function onActiveCellChanged(): void {
+    if (keyMap === 'vim') {
+      activeCell = notebookTracker.activeCell;
+      if (activeCell) {
+        // From https://github.com/axelfahy/jupyterlab-vim/blob/c4e43f940ef4be4c961608b6192412d2f3a33d1f/src/index.ts
+        CodeMirror.prototype.save = () => {
+          void commands.execute('docmanager:save');
+        };
+        const lvim = (CodeMirror as any).Vim as any;
+        if (lvim) {
+          lvim.defineEx('quit', 'q', function(cm: any) {
+            void commands.execute('notebook:enter-command-mode');
+          });
+          lvim.defineAction('splitCell', (cm: any, actionArgs: any) => {
+            void commands.execute('notebook:split-cell-at-cursor');
+          });
+          lvim.mapCommand('-', 'action', 'splitCell', {}, { extra: 'normal' });
+        }
+      }
+    }
+  }
+
```

Comment:
I think all of these act globally and thus only need to be run when the vim keymap is selected rather than everytime the active cell is changed. I'd suggest moving these definitions under the `if( keymap == 'vim')`.